src/sentry/migrations/0181_copy_useroptions_to_notificationsettings.py

```diff
@@ -0,0 +1,152 @@
+        elif user_option.key == "mail:alert":  # fine tuned project alerts
+            if user_option.project_id:
+                scope_identifier = user_option.project_id
+                scope_type = NotificationScopeType.PROJECT.value
+            else:
+                scope_identifier = user_option.user.id
+                scope_type = NotificationScopeType.USER.value
+            type = NotificationSettingTypes.ISSUE_ALERTS.value
+            if int(user_option.value) == 0:
+                value = NotificationSettingOptionValues.NEVER.value
+            if int(user_option.value) == 1:
+                value = NotificationSettingOptionValues.ALWAYS.value
+        elif user_option.key == "subscribe_by_default":  # top level project alerts on/off
+            scope_identifier = user_option.user.id
+            scope_type = NotificationScopeType.USER.value
+            type = NotificationSettingTypes.ISSUE_ALERTS.value
+            if int(user_option.value) == 1:
+                value = NotificationSettingOptionValues.ALWAYS.value
+            if int(user_option.value) == 0:
+                value = NotificationSettingOptionValues.NEVER.value
+        elif user_option.key == "deploy-emails":
+            # if you have fine tuning for an org, organization_id will have a value (rather than None)
+            if user_option.organization_id:
+                scope_identifier = user_option.organization_id
+                scope_type = NotificationScopeType.ORGANIZATION.value
+            else:
+                scope_identifier = user_option.user.id
+                scope_type = NotificationScopeType.USER.value
+            type = NotificationSettingTypes.DEPLOY.value
+            # if you've not explicitly set anything OR set it to default, there is no db row
+            # by default deploy notifications are set to committed_deploys_only,
+            # but there will be an entry for the top level alert option
+            # if you change the value to something else
+            if user_option.value == UserOptionValue.all_deploys:
+                value = NotificationSettingOptionValues.ALWAYS.value
+            if user_option.value == UserOptionValue.no_deploys:
+                value = NotificationSettingOptionValues.NEVER.value
+            if user_option.value == UserOptionValue.committed_deploys_only:
+                value = NotificationSettingOptionValues.COMMITTED_ONLY.value
+        else:
+            continue
+        user = User.objects.select_related("actor").get(id=user_option.user_id)
+        NotificationSetting.objects.update_or_create(
+            scope_type=scope_type,  # user, org, or project
+            scope_identifier=scope_identifier,  # user_id, organization_id, or project_id
+            target=user.actor,
+            provider=ExternalProviders.EMAIL.value,  # 100
+            type=type,
+            value=value,  # NotificationSettingOptionValues
+        )
```

Comment:
When you use `update_or_create`, the kwargs passed directly to the function are used to check for an existing row, and if it doesn't exist then they're used to create a new row. There's also an optional `defaults` kwarg which takes a dict of values to be used on an update (they will also be used on a create).

What you probably want to do here is have `value` passed in via `defaults`. So your lookup ends up being on the five values that make up your unique key. This avoids a problem where if a row exists with a different `value` it would attempt to create a new row and fail.

I think that in your case actually this is fine as is and won't cause problems, but people tend to copy/paste example code from migrations into future ones, and so it's better to use it properly.